File: src/test/scala/WrapperTypeTest.scala
/* -*-mode:scala; c-basic-offset:2; indent-tabs-mode:nil -*- */
package com.jcraft.jsch.jzlib

import org.scalatest._
import org.scalatest.flatspec._
import matchers.should._

import scala.language.reflectiveCalls

import java.io.{ByteArrayOutputStream => BAOS, ByteArrayInputStream => BAIS}

import JZlib._

class WrapperTypeTest extends AnyFlatSpec with BeforeAndAfter with Matchers {
  private val data = "hello, hello!".getBytes

  private val comprLen = 40000
  private val uncomprLen = comprLen
  private var compr:Array[Byte] = _
  private var uncompr:Array[Byte] = _
  private var err: Int = _

  private val cases = /* success */     /* fail */ 
    List((W_ZLIB, (List(W_ZLIB, W_ANY), List(W_GZIP, W_NONE))),
         (W_GZIP, (List(W_GZIP, W_ANY), List(W_ZLIB, W_NONE))),
         (W_NONE, (List(W_NONE, W_ANY), List(W_ZLIB, W_GZIP))))

  before {
    compr = new Array[Byte](comprLen)
    uncompr = new Array[Byte](uncomprLen)

    err = Z_OK
  }

  after {
  }

  behavior of "Deflater"

  it can "detect data type of input." in {
    implicit val buf = compr

    cases foreach { case (iflag, (good, bad)) => 
      val baos = new BAOS
      val deflater = new Deflater(Z_DEFAULT_COMPRESSION, DEF_WBITS, 9, iflag)
      val gos = new DeflaterOutputStream(baos, deflater)
      data -> gos
      gos.close

      val deflated = baos.toByteArray

      good map { w =>
        val baos2 = new BAOS
        val inflater = new Inflater(w)
        new InflaterInputStream(new BAIS(deflated), inflater) -> baos2
        val data1 = baos2.toByteArray
        data1.length should equal (data.length)
        data1 should equal (data)
        import inflater._
        (avail_in, avail_out, total_in, total_out)
      } reduceLeft { (x, y) => x should equal (y); x }

      bad foreach { w =>
        val baos2 = new BAOS
        val inflater = new Inflater(w)
        try {
          new InflaterInputStream(new BAIS(deflated), inflater) -> baos2
          fail("unreachable")
        }
        catch {
          case e:java.io.IOException  =>
        }
      } 
    }
  } 

  behavior of "ZStream"

  it can "detect data type of input." in {
    cases foreach { case (iflag, (good, bad)) => 
      val deflater = new ZStream

      err = deflater.deflateInit(Z_BEST_SPEED, DEF_WBITS, 9, iflag)
      err should equal (Z_OK)

      deflate(deflater, data, compr)

      good foreach { w =>
        val inflater = inflate(compr, uncompr, w)
        val total_out = inflater.total_out.asInstanceOf[Int]
        new String(uncompr, 0, total_out) should equal (new String(data))
      }

      bad foreach { w =>
        inflate_fail(compr, uncompr, w)
      }
    }
  }

  behavior of "Deflater"

  it should "support wbits+32." in {

    var deflater = new Deflater
    err = deflater.init(Z_BEST_SPEED, DEF_WBITS, 9)
    err should equal (Z_OK)

    deflate(deflater, data, compr)

    var inflater = new Inflater
    err = inflater.init(DEF_WBITS + 32)
    err should equal (Z_OK)

    inflater.setInput(compr)

    var loop = true
    while(loop) {
      inflater.setOutput(uncompr)
      err = inflater.inflate(Z_NO_FLUSH)
      if(err == Z_STREAM_END) loop = false
      else err should equal (Z_OK)
    }
    err = inflater.end
    err should equal (Z_OK)

    var total_out = inflater.total_out.asInstanceOf[Int]
    new String(uncompr, 0, total_out) should equal (new String(data))

    deflater = new Deflater
    err = deflater.init(Z_BEST_SPEED, DEF_WBITS + 16, 9)
    err should equal (Z_OK)

    deflate(deflater, data, compr)

    inflater = new Inflater
    err = inflater.init(DEF_WBITS + 32)
    err should equal (Z_OK)

    inflater.setInput(compr)

    loop = true
    while(loop) {
      inflater.setOutput(uncompr)
      err = inflater.inflate(Z_NO_FLUSH)
      if(err == Z_STREAM_END) loop = false
      else err should equal (Z_OK)
    }
    err = inflater.end
    err should equal (Z_OK)

    total_out = inflater.total_out.asInstanceOf[Int]
    new String(uncompr, 0, total_out) should equal (new String(data))
  }

  private def deflate(deflater: ZStream,
                      data: Array[Byte], compr: Array[Byte]) = {
    deflater.setInput(data)
    deflater.setOutput(compr)

    err = deflater.deflate(JZlib.Z_FINISH)
    err should equal (Z_STREAM_END)

    err = deflater.end
    err should equal (Z_OK)
  }    

  private def inflate(compr: Array[Byte],
                      uncompr: Array[Byte],
                      w: WrapperType) = {
    val inflater = new ZStream
    err = inflater.inflateInit(w)
    err should equal (Z_OK)

    inflater.setInput(compr)

    var loop = true
    while(loop) {
      inflater.setOutput(uncompr)
      err = inflater.inflate(Z_NO_FLUSH)
      if(err == Z_STREAM_END) loop = false
      else err should equal (Z_OK)
    }
    err = inflater.end
    err should equal (Z_OK)

    inflater
  }

  private def inflate_fail(compr: Array[Byte],
                           uncompr: Array[Byte],
                           w: WrapperType) = {
    val inflater = new ZStream

    err = inflater.inflateInit(w)
    err should equal (Z_OK)

    inflater.setInput(compr)

    var loop = true
    while(loop) {
      inflater.setOutput(uncompr)
      err = inflater.inflate(Z_NO_FLUSH)
      if(err == Z_STREAM_END) loop = false
      else {
        err should equal (Z_DATA_ERROR)
        loop = false
      }
    }
  }
}

File: src/test/java/com/jcraft/jsch/jzlib/WrapperTypeTest.java
package com.jcraft.jsch.jzlib;

import static com.jcraft.jsch.jzlib.JZlib.*;
import static com.jcraft.jsch.jzlib.Package.*;
import static java.nio.charset.StandardCharsets.UTF_8;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class WrapperTypeTest {
  private final byte[] data = "hello, hello!".getBytes(UTF_8);

  private final int comprLen = 40000;
  private final int uncomprLen = comprLen;
  private byte[] compr;
  private byte[] uncompr;
  private int err;

  private final List<Case> cases = Arrays.asList(
      /* success fail */
      new Case(W_ZLIB, Arrays.asList(W_ZLIB, W_ANY), Arrays.asList(W_GZIP, W_NONE)),
      new Case(W_GZIP, Arrays.asList(W_GZIP, W_ANY), Arrays.asList(W_ZLIB, W_NONE)),
      new Case(W_NONE, Arrays.asList(W_NONE, W_ANY), Arrays.asList(W_ZLIB, W_GZIP)));

  @BeforeEach
  public void before() {
    compr = new byte[comprLen];
    uncompr = new byte[uncomprLen];

    err = Z_OK;
  }

  @AfterEach
  public void after() {}

  @Test
  public void testDeflaterCanDetectDataTypeOfInput() {
    byte[] buf = compr;

    cases.forEach(uncheckedConsumer(c -> {
      ByteArrayOutputStream baos = new ByteArrayOutputStream();
      Deflater deflater = new Deflater(Z_DEFAULT_COMPRESSION, DEF_WBITS, 9, c.iflag);
      DeflaterOutputStream gos = new DeflaterOutputStream(baos, deflater);
      readArray(data, gos, buf);
      gos.close();

      byte[] deflated = baos.toByteArray();

      c.good.stream().map(uncheckedFunction(w -> {
        ByteArrayOutputStream baos2 = new ByteArrayOutputStream();
        Inflater inflater = new Inflater(w);
        readIS(new InflaterInputStream(new ByteArrayInputStream(deflated), inflater), baos2, buf);
        byte[] data1 = baos2.toByteArray();
        assertEquals(data.length, data1.length);
        assertArrayEquals(data, data1);
        return new Tuple(inflater.avail_in, inflater.avail_out, inflater.total_in,
            inflater.total_out);
      })).reduce((x, y) -> {
        assertEquals(y, x);
        return x;
      });

      c.bad.forEach(uncheckedConsumer(w -> {
        ByteArrayOutputStream baos2 = new ByteArrayOutputStream();
        Inflater inflater = new Inflater(w);
        assertThrows(IOException.class,
            () -> readIS(new InflaterInputStream(new ByteArrayInputStream(deflated), inflater),
                baos2, buf));
      }));
    }));
  }

  @Test
  public void testZStreamCanDetectDataTypeOfInput() {
    cases.forEach(c -> {
      ZStream deflater = new ZStream();

      err = deflater.deflateInit(Z_BEST_SPEED, DEF_WBITS, 9, c.iflag);
      assertEquals(Z_OK, err);

      deflate(deflater, data, compr);

      c.good.forEach(w -> {
        ZStream inflater = inflate(compr, uncompr, w);
        int total_out = (int) inflater.total_out;
        assertEquals(new String(data, UTF_8), new String(uncompr, 0, total_out, UTF_8));
      });

      c.bad.forEach(w -> {
        inflate_fail(compr, uncompr, w);
      });
    });
  }

  @Test
  public void testDeflaterCanSupportWbitsPlus32() {

    Deflater deflater = new Deflater();
    err = deflater.init(Z_BEST_SPEED, DEF_WBITS, 9);
    assertEquals(Z_OK, err);

    deflate(deflater, data, compr);

    Inflater inflater = new Inflater();
    err = inflater.init(DEF_WBITS + 32);
    assertEquals(Z_OK, err);

    inflater.setInput(compr);

    boolean loop = true;
    while (loop) {
      inflater.setOutput(uncompr);
      err = inflater.inflate(Z_NO_FLUSH);
      if (err == Z_STREAM_END)
        loop = false;
      else
        assertEquals(Z_OK, err);
    }
    err = inflater.end();
    assertEquals(Z_OK, err);

    int total_out = (int) inflater.total_out;
    assertEquals(new String(data, UTF_8), new String(uncompr, 0, total_out, UTF_8));

    deflater = new Deflater();
    err = deflater.init(Z_BEST_SPEED, DEF_WBITS + 16, 9);
    assertEquals(Z_OK, err);

    deflate(deflater, data, compr);

    inflater = new Inflater();
    err = inflater.init(DEF_WBITS + 32);
    assertEquals(Z_OK, err);

    inflater.setInput(compr);

    loop = true;
    while (loop) {
      inflater.setOutput(uncompr);
      err = inflater.inflate(Z_NO_FLUSH);
      if (err == Z_STREAM_END)
        loop = false;
      else
        assertEquals(Z_OK, err);
    }
    err = inflater.end();
    assertEquals(Z_OK, err);

    total_out = (int) inflater.total_out;
    assertEquals(new String(data, UTF_8), new String(uncompr, 0, total_out, UTF_8));
  }

  private void deflate(ZStream deflater, byte[] data, byte[] compr) {
    deflater.setInput(data);
    deflater.setOutput(compr);

    err = deflater.deflate(JZlib.Z_FINISH);
    assertEquals(Z_STREAM_END, err);

    err = deflater.end();
    assertEquals(Z_OK, err);
  }

  private ZStream inflate(byte[] compr, byte[] uncompr, WrapperType w) {
    ZStream inflater = new ZStream();
    err = inflater.inflateInit(w);
    assertEquals(Z_OK, err);

    inflater.setInput(compr);

    boolean loop = true;
    while (loop) {
      inflater.setOutput(uncompr);
      err = inflater.inflate(Z_NO_FLUSH);
      if (err == Z_STREAM_END)
        loop = false;
      else
        assertEquals(Z_OK, err);
    }
    err = inflater.end();
    assertEquals(Z_OK, err);

    return inflater;
  }

  private void inflate_fail(byte[] compr, byte[] uncompr, WrapperType w) {
    ZStream inflater = new ZStream();

    err = inflater.inflateInit(w);
    assertEquals(Z_OK, err);

    inflater.setInput(compr);

    boolean loop = true;
    while (loop) {
      inflater.setOutput(uncompr);
      err = inflater.inflate(Z_NO_FLUSH);
      if (err == Z_STREAM_END)
        loop = false;
      else {
        assertEquals(Z_DATA_ERROR, err);
        loop = false;
      }
    }
  }

  static class Case {
    final WrapperType iflag;
    final List<WrapperType> good;
    final List<WrapperType> bad;

    Case(WrapperType iflag, List<WrapperType> good, List<WrapperType> bad) {
      this.iflag = iflag;
      this.good = good;
      this.bad = bad;
    }
  }

  static class Tuple {
    private final int a;
    private final int b;
    private final long c;
    private final long d;

    Tuple(int a, int b, long c, long d) {
      this.a = a;
      this.b = b;
      this.c = c;
      this.d = d;
    }

    @Override
    public boolean equals(Object obj) {
      if (!(obj instanceof Tuple))
        return false;
      else if (a != ((Tuple) obj).a)
        return false;
      else if (b != ((Tuple) obj).b)
        return false;
      else if (c != ((Tuple) obj).c)
        return false;
      else if (d != ((Tuple) obj).d)
        return false;
      else
        return true;
    }

    @Override
    public int hashCode() {
      return Objects.hash(a, b, c, d);
    }
  }
}
